syntax = "proto3";

package preprocessing;

import "google/protobuf/timestamp.proto";

// As defined in
// https://eepublicdownloads.entsoe.eu/clean-documents/Publications/SOC/Continental_Europe/
// 150420_quality_of_datasets_and_calculations_3rd_edition.pdf
enum UCTERegionType {
    A = 0;
    B = 1;
    C = 2;
    D = 3;
    D1 = 4;
    D2 = 5;
    D4 = 6;
    D6 = 7;
    D7 = 8;
    D8 = 9;
    E = 10;
    F = 11;
    G = 12;
    H = 13;
    I = 14;
    J = 15;
    K = 16;
    L = 17;
    M = 18;
    N = 19;
    O = 20;
    P = 21;
    Q = 22;
    R = 23;
    S = 24;
    T = 25;
    U = 26;
    V = 27;
    W = 28;
    X = 29;
    Y = 30;
    Z = 31;
    ZERO = 32;
    TWO = 33;
    UNDERSCORE = 34;
}

enum CGMESRegionType {
    AL = 0;
    AD = 1;
    AM = 2;
    AT = 3;
    AZ = 4;
    BY = 5;
    BE = 6;
    BA = 7;
    BG = 8;
    HR = 9;
    CY = 10;
    CZ = 11;
    DK = 12;
    EE = 13;
    FI = 14;
    FR = 15;
    GE = 16;
    DE = 17;
    GR = 18;
    HU = 19;
    IS = 20;
    IE = 21;
    IT = 22;
    LV = 23;
    LI = 24;
    LT = 25;
    LU = 26;
    MT = 27;
    MD = 28;
    MC = 29;
    ME = 30;
    NL = 31;
    MK = 32;
    NO = 33;
    PL = 34;
    PT = 35;
    RO = 36;
    RU = 37;
    SM = 38;
    RS = 39;
    SK = 40;
    SI = 41;
    ES = 42;
    SE = 43;
    CH = 44;
    TR = 45;
    UA = 46;
    GB = 47;
    VA = 48;
}

enum GridModelType {
    UCTE = 0;
    CGMES = 1;
}

enum RealiseStationBusbarChoiceHeuristic {
    FIRST = 0;
    LEAST_CONNECTED_BUSBAR = 1;
}


// ---------------------------
// LimitAdjustmentParameters
// ---------------------------
message LimitAdjustmentParameters {
    double n_0_factor = 1;
    double n_1_factor = 2;
    double n_0_min_increase = 3;
    double n_1_min_increase = 4;
}

// ---------------------------
// AreaSettings
// ---------------------------
message AreaSettings {
    oneof{
        repeated CGMESRegionType control_area = 1;
        repeated UCTERegionType control_area = 2;
    }
    oneof{
        repeated CGMESRegionType view_area = 9;
        repeated UCTERegionType view_area = 10;
    }

    oneof{
        repeated CGMESRegionType nminus1_area = 9;
        repeated UCTERegionType nminus1_area = 10;
    }
    uint32 cutoff_voltage = 4;
    LimitAdjustmentParameters dso_trafo_factors = 5;
    double dso_trafo_weight = 6;
    LimitAdjustmentParameters border_line_factors = 7;
    double border_line_weight = 8;
}

message BaseImporterParameters {
    AreaSettings area_settings = 1;
    string data_folder = 2;
    string grid_model_file = 3;
    GridModelType data_type = 4;
    string white_list_file = 5;
    string black_list_file = 6;
    string ignore_list_file = 7;
    string ingress_id = 8;
    string contingency_list_file = 9;
    string schema_format = 10;
}

message PreprocessParameters {
    int32 filter_disconnectable_branches_processes = 1;
    bool action_set_filter_bridge_lookup = 2;
    bool action_set_filter_bsdf_lodf = 3;
    int32 action_set_filter_bsdf_lodf_batch_size = 4;
    int32 action_set_clip = 5;
    bool asset_topo_close_couplers = 6;
    RealiseStationBusbarChoiceHeuristic realise_station_busbar_choice_heuristic = 7;
    double ac_dc_interpolation = 8;
    bool enable_n_2 = 9;
    double n_2_more_splits_penalty = 10;
    bool enable_bb_outage = 11;
    bool bb_outage_as_nminus1 = 12;
    double bb_outage_more_splits_penalty = 13;
    bool clip_bb_outage_penalty = 14;
    double double_limit_n0 = 15;
    double double_limit_n1 = 16;
    int32 initial_loadflow_processes = 17;
}

message StartPreprocessingCommand {
    oneof importer_parameters {
        UcteImporterParameters ucte_importer_parameters = 1;
        CgmesImporterParameters cgmes_importer_parameters = 2;
    }
    PreprocessParameters preprocess_parameters = 3;
    string preprocess_id = 4;
}

message ShutdownCommand {
    int32 exit_code = 1;
}

message Command {
    oneof command {
        StartPreprocessingCommand start_preprocessing = 1;
        ShutdownCommand shutdown = 2;
    }
    google.protobuf.Timestamp timestamp = 3;
    string uuid = 4;
}
